src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -542,32 +554,45 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
         }
     }
 
-    private void FinishWritingHeaders(int streamId, int payloadLength, bool done)
+    private void SplitHeaderFramesToOutput(int streamId, HeaderWriteResult done, bool isFramePrepared)
     {
-        var buffer = _headerEncodingBuffer.AsSpan();
-        _outgoingFrame.PayloadLength = payloadLength;
-        if (done)
+        var dataToFrame = _headerEncodingBuffer.WrittenSpan;
+        var shouldPrepareFrame = !isFramePrepared;
+        while (dataToFrame.Length > 0)
         {
-            _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
-        }
-
-        WriteHeaderUnsynchronized();
-        _outputWriter.Write(buffer.Slice(0, payloadLength));
-
-        while (!done)
-        {
-            _outgoingFrame.PrepareContinuation(Http2ContinuationFrameFlags.NONE, streamId);
-
-            done = HPackHeaderWriter.ContinueEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
-            _outgoingFrame.PayloadLength = payloadLength;
+            if (shouldPrepareFrame)
+            {
+                _outgoingFrame.PrepareContinuation(Http2ContinuationFrameFlags.NONE, streamId);
+            }
+            else
+            {
+                shouldPrepareFrame = true;
+            }
 
-            if (done)
+            var currentSize = dataToFrame.Length > _maxFrameSize ? _maxFrameSize : dataToFrame.Length;
```

Comment:
please check my logic, but: `= Math.Min(dataToFrame.Length, _maxFrameSize)` ?

Reply:
I can do that, I was not sure  if it is not worse perf wise.

Reply:
this is going to be irrelevant in terms of perf; hypothetically `Min`/`Max` *could be infinitesimally faster (if they can use an intrinsic rather than a branch), but in reality they'll be identical